### PR TITLE
Feat: contact us endpoint for users

### PR DESCRIPTION
**Pull request by Xaldovah**
## Description
This PR introduces a new API endpoint to handle user inquiries submitted through the "Contact Us" form. The endpoint validates the input data, sends a request to an internal email service endpoint, and returns appropriate responses based on the success or failure of the operation.

## Linked Issue
This PR addresses the following issue: [Issue #101](https://github.com/hngprojects/hng_boilerplate_nestjs/issues/101)

## Changes Made
1. **API Endpoint Implementation:**
   - Added a POST endpoint `/api/v1/contact` to accept inquiries.
   - The endpoint validates the incoming request payload to ensure all required fields (`name`, `email`, `message`) are present and valid.

2. **Data Validation and Sanitization:**
   - Used Pydantic models to validate the request payload:
     - `name`: Non-empty string.
     - `email`: Valid email address.
     - `message`: Non-empty string.

3. **Email Sending:**
   - Integrated with an internal email service by simulating email sending functionality within the app.

4. **Response Handling:**
   - On successful request to the email service, returns a 200 OK status with a success message.
   - On failure, returns a 500 Internal Server Error with an appropriate error message.

5. **Logging:**
   - Set up logging to capture the flow of operations and errors for better monitoring and debugging.
   - Logs are written to `logs/contact_us.log` with a specific format.

6. **Testing:**
   - Added unit tests to cover various scenarios:
     - Successful inquiry submission.
     - Email sending failure.
     - Invalid email format.
     - Missing required fields.
     - Empty message.
   - Performed load testing using k6 to ensure the endpoint can handle a high volume of successful requests.

## Load Testing Results
Performed load testing with k6 to verify the endpoint can handle multiple concurrent requests efficiently. The test simulated 50 virtual users for a duration of 5 minutes. The results indicated that the endpoint can handle the load, with the majority of requests being processed successfully.

## Files Changed
1. `api/v1/routes/contact_us.py`:
   - Implemented the API endpoint and logic to send a request to the internal email service.
   - Added logging for better error tracking and monitoring.

2. `tests/test_contact_us.py`:
   - Added unit tests to cover various scenarios for the "Contact Us" endpoint.

## Documentation
**API Documentation:**

- **Endpoint:** `/api/v1/contact`
- **Method:** POST
- **Request Body:**
  - `name`: (string) The name of the user.
  - `email`: (string) The email address of the user.
  - `message`: (string) The message from the user.

**Example Request:**

_http_
_POST /api/v1/contact_
_Content-Type: application/json_
```
{
  "name": "John Doe",
  "email": "john.doe@example.com",
  "message": "Hello, I have an inquiry."
}

```
**Successful Response:**

```
{
  "message": "Inquiry sent successfully",
  "status": 200
}
```

**Error Response (Invalid Email):**

```
{
  "detail": [
    {
      "loc": ["body", "email"],
      "msg": "value is not a valid email address",
      "type": "value_error.email"
    }
  ]
}

```

**Error Response (Email Sending Failure):**

```
{
  "detail": "Failed to send inquiry"
}
```

#### Review and Merge
You can kindly review the changes that I have made and if there are any actions I should take, let me know. If you are satisfied with this PR you can merge it.

### CODE REVIEW

**Comment by Dev-wonderful** [comment]
Resolve your conflicts

**Comment by Xaldovah** [comment]
@Dev-wonderful I have resolved the conflicts now.

File: api/v1/routes/contact_us.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+
+class ContactForm(BaseModel):
+    name: constr(strip_whitespace=True, min_length=1)
+    email: EmailStr
+    message: constr(strip_whitespace=True, min_length=1)
+
+def send_email_via_service(name: str, email: str, message: str) -> bool:
+    url = "http://localhost:8000/api/v1/send-mail"
+    headers = {
```
Comment (Dev-wonderful) [inline api/v1/routes/contact_us.py:21]: The `url` is not needed, emailing functionality would be a service within the app

File: api/v1/routes/contact_us.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+from pydantic import BaseModel, EmailStr, constr
+from fastapi.responses import JSONResponse
+import requests
+
+
+router = APIRouter()
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
```
Comment (Dev-wonderful) [inline api/v1/routes/contact_us.py:12]: setup the logger properly. Create a logging folder, set the name of the log files, set the format.

Also the logging level doesn't mean only `INFO` would be logged, just incase you don't know.

File: scripts/load_test.js
Comment (Dev-wonderful) [inline scripts/load_test.js:1]: This should remain in your local dev environment

**Review by Dev-wonderful — CHANGES_REQUESTED** [review]

File: api/v1/routes/contact_us.py (hunk 1 above)
Comment (Xaldovah) [reply api/v1/routes/contact_us.py]: I have implemented like you advised. I have presumed that the email service file is in the services directory and I have used the following code as my email_services to simulate that my code is working properly.

```
import os
import requests
import logging
from dotenv import load_dotenv

load_dotenv()

# Setup logging
logger = logging.getLogger(__name__)


class EmailService:
    def __init__(self):
        self.api_key = os.getenv('BREVO_API_KEY')
        self.receiver_email = os.getenv('RECEIVER_EMAIL')

    def send_email(self, name: str, email: str, message: str) -> bool:
        url = "https://api.brevo.com/v3/smtp/email"
        headers = {
                "accept": "application/json",
                "api-key": self.api_key,
                "content-type": "application/json"
        }
        data = {
                "sender": {"name": name, "email": email},
                "to": [{"email": self.receiver_email}],
                "subject": "New Contact Us Inquiry",
                "htmlContent": f"<p>Name: {name}</p><p>Email: {email}</p><p>Message: {message}</p>"
        }

        try:
            response = requests.post(url, json=data, headers=headers)
            response.raise_for_status()
            logger.info(f"Email sent successfully to {self.receiver_email}")
            return True
        except requests.exceptions.HTTPError as err:
            logger.error(f"Failed to send email: {err}")
            return False
```

File: api/v1/routes/contact_us.py (hunk 2 above)
Comment (Xaldovah) [reply api/v1/routes/contact_us.py]: I have enhanced the code to make sure it creates a log file within a directory and improved the logging system to log more details.

File: scripts/load_test.js
Comment (Xaldovah) [reply scripts/load_test.js:1]: I have deleted this file from the files that I have pushed.

**Comment by Xaldovah** [comment]
@Dev-wonderful I have updated my code to follow the advised instructions.

**Comment by Xaldovah** [comment]
Since I don't know where and how the internal email service will be and be implemented, I have presumed that the email service file is in the services directory and I have used the following code as my email_services to simulate that the code is working properly.

```
import os
import requests
import logging
from dotenv import load_dotenv

load_dotenv()

# Setup logging
logger = logging.getLogger(__name__)


class EmailService:
    def __init__(self):
        self.api_key = os.getenv('BREVO_API_KEY')
        self.receiver_email = os.getenv('RECEIVER_EMAIL')

    def send_email(self, name: str, email: str, message: str) -> bool:
        url = "https://api.brevo.com/v3/smtp/email"
        headers = {
                "accept": "application/json",
                "api-key": self.api_key,
                "content-type": "application/json"
        }
        data = {
                "sender": {"name": name, "email": email},
                "to": [{"email": self.receiver_email}],
                "subject": "New Contact Us Inquiry",
                "htmlContent": f"<p>Name: {name}</p><p>Email: {email}</p><p>Message: {message}</p>"
        }

        try:
            response = requests.post(url, json=data, headers=headers)
            response.raise_for_status()
            logger.info(f"Email sent successfully to {self.receiver_email}")
            return True
        except requests.exceptions.HTTPError as err:
            logger.error(f"Failed to send email: {err}")
            return False
```

**Comment by Xaldovah** [comment]
Removed the test file because the previous ones that were on the dev branch were removed to resolve test issues

**Pull request closed without merging**